### PR TITLE
Enable deletion of rejected expenses

### PR DESCRIPTION
**Pull request by flickz**
Resolves https://github.com/opencollective/opencollective/issues/2337

TODO:

- [x] Improve the reject button on hover

- [x] Add e2e test

delete expense

### CODE REVIEW

**Comment by znarf** [comment]
Good progress. Can you share a screen record of the feature?

I would also harden `deleteExpense` on the API side to make sure we can only delete rejected expenses. What do you think?

**Comment by flickz** [comment]
Yes we should ensure that.

File: components/expenses/Expense.js [hunk 1]
```diff
@@ -23,6 +25,14 @@ import EditPayExpenseFeesForm from './EditPayExpenseFeesForm';
...
 import StyledButton from '../StyledButton';
 
+const DeleteButton = styled(StyledButton)`
+  border: ${themeGet('colors.red.600')};
+  &:hover {
+    opacity: 0.8;
+    background: ${themeGet('colors.red.600')};
+  }
+`;
+
```
Comment (Betree) [inline components/expenses/Expense.js]: A state for `danger` button style has been added to: https://www.figma.com/file/N4Xbl652BhzOutXmzhcrLGch/DS-06-%2F-Atoms?node-id=1928%3A7979. Let's add this to `lib/theme/index.js` and use the standard `StyledButton`.

image

File: components/expenses/Expense.js [hunk 2]
```diff
@@ -144,11 +164,22 @@ class Expense extends React.Component {
...
-      this.setState({ showUnapproveModal: false });
+      const error = getErrorFromGraphqlException(err);
+      this.setState({ showUnapproveModal: false, error: error.message });
+    }
+  };
+
+  handleDeleteExpense = async id => {
+    try {
+      await this.props.deleteExpense(id);
+      this.setState({ showUnapproveModal: false, error: null });
```
Comment (Betree) [inline components/expenses/Expense.js]: Shouldn't that be

```suggestion
      this.setState({ showDeleteExpenseModal: false, error: null });
```

File: components/expenses/Expense.js [hunk 3]
```diff
@@ -109,6 +121,14 @@ class Expense extends React.Component {
       },
       no: { id: 'no', defaultMessage: 'No' },
       yes: { id: 'yes', defaultMessage: 'Yes' },
+      'delete.modal.header': {
+        id: 'delete.modal.header',
```
Comment (Betree) [inline components/expenses/Expense.js]: Id should not be `delete.modal.header` if this is about deleting an expense.

```suggestion
        id: 'deleteExpense.modal.header',
```

File: components/expenses/Expense.js [hunk 4]
```diff
@@ -109,6 +121,14 @@ class Expense extends React.Component {
...
+        defaultMessage: 'Delete Expense',
+      },
+      'delete.modal.body': {
+        id: 'delete.modal.body',
```
Comment (Betree) [inline components/expenses/Expense.js]: Id should not be `delete.modal.body` if this is about deleting an expense.

```suggestion
        id: 'deleteExpense.modal.body',
```

File: components/expenses/Expense.js [hunk 5]
```diff
@@ -468,17 +516,49 @@ class Expense extends React.Component {
...
+                        <FormattedMessage id="expense.delete.btn" defaultMessage="Delete" />
+                      </DeleteButton>
+                    )}
                   </div>
                 </div>
               )}
             </div>
           )}
+          {this.state.error && (
+            <div className="error" data-cy="errorMessage">
```
Comment (Betree) [inline components/expenses/Expense.js]: We should prefer styled-components for new developments. This one can easily be replaced by a 
```suggestion
            <P color="red.500" data-cy="errorMessage">
```

File: components/expenses/Expense.js [hunk 6]
```diff
@@ -23,6 +24,23 @@ import EditPayExpenseFeesForm from './EditPayExpenseFeesForm';
...
+
+  &:focus {
+    background: #f51d57;
+  }
+
+  &:hover {
+    background: #ff4778;
+  }
+`;
+
```
Comment (Betree) [inline components/expenses/Expense.js]: Can't we rather use the `buttonStyle="danger"`?

**Comment by flickz** [comment]
You can't approve a rejected expense after a specific period probably why.
https://github.com/opencollective/opencollective-frontend/blob/master/components/expenses/Expense.js#L208

I'm not sure why it is slow from your end but maybe instead of using `destory()` in the api we can just update the expense status to `DELETED` but I don't think soft delete should be slow.

**Comment by znarf** [comment]
@flickz rebased to fix conflict and we can merge

**Pull request merged**